\section{Introduction} 
As sequencing technologies improve, the analysis of microbial community composition and function has rapidly advanced by focusing on a small number of phylogenetically informative genes surveys utilizing loci such as the small subunit of ribosomal RNA (SSU rRNA). The ability to profile community structures provides new insights into the role of microbiomes in human health, soil ecology and environmental remediation (REFS). Nevertheless, the single gene survey approach has limited taxonomic resolution because microorganisms with significantly different genomic functionalities can be incorrectly clustered together. 

Functional profiling based on Whole Metagenome Shotgun (WMS) data strives to catalog the genes present in a community. However, the utility of random DNA sequence reads from metagenome is limited by the incomplete nature of annotated reference databases. Specifically, methods for rapidly mapping DNA reads to annotated reference genes fail when the references within the curated databases moderately diverged from the sequences of homologous genes in metagenomics samples.  The dramatic increase in number and length of high quality sequences (hundreds of millions of reads with lengths routinely reaching 250 bases) has the potential to provide evidence of genes over a large dynamic range. Additionally, these improvements allow for much greater sensitivity in true metagenomic samples where many coding sequences are low abundance and unlikely to assemble.  

To improve the mapping of DNA sequence reads to homologous protein coding DNA sequences in reference databases, we propose a mapping technique that takes into account higher divergence rates at silent sites in coding sequences due to purifying selection in comparison to nucleotide positions where a mutation will result in amino acid changes. Consequently, read mapping that exploits constraints of protein coding function and disregards the rapidly diverging positions has the potential to significantly increase efficiency of matching reads to protein coding regions of a reference without compromising accuracy. This kind of mapping can be implemented in multiple ways. First, we could map DNA reads to DNA references by neglecting any differences at silent sites in the coding sequences; Novalign utilizes this mapping mode but has yet to be employed in this context. Alternatively, we may also map the putative amino acid sequences encoded by WGS reads directly to reference protein sequences. To our knowledge, no method exists that directly maps amino acid sequences to reference proteins. Here, we introduce PALADIN, a modified version of the popular BWA mapping algorithm that maps the translations of WMS reads to protein references in a similar manner as BWA. 

For the three experiments conducted, we chose to harness the Uniprot database as a reference and mined DNA sequences corresponding to each protein entry. We also derived a mock metagenomic data set of six standard well-annotated reference genomes with diverse gene contents and base composition.  Reads were generated using ART (reference!). 
First, we mapped these reads to the nucleotide sequences corresponding to the reference protein sequences in the Uniprot database (Huang et al., 2012) using BWA (Li and Durbin, 2009) and Novoalign (Novocraft.com) (figure 1A). In this case the maximum proportion of the reads that mapped was 22.1% (BWA)  and 12.1% (Novoalign).  Since the primary goal of this approach is to define the potential coding functions in WMS read data, we evaluated the success based on the efficiency (percent of reads mapped), accuracy and sensitivity of the functional match (based on the degree of similarity in GO terms) between the source of the read and the protein to which that read mapped  (see supplemental methods). 

In the second experiment, we used Novoalign to map the same reads to the same coding sequence references, but we converted the reads into their degenerate sequences by replacing silent nucleotides with their degenerate bases as indicated by ambiguous IUPAC codes. Novalign is one of the few currently available algorithms that allows read mapping to degenerate references. The same methods were used to evaluate accuracy and sensitivity as the BWA mapping technique. TONI WILL GET THE NUMBERS TO ME!  Describe results.
Finally, we executed a new algorithm, the Protein Alignment and Detection Interface (PALADIN), that extends the highly efficient and well-tested Burrows-Wheeler Alignment (BWA) codebase to operate in protein space.  In this process potential proteins encoded by the reads are mapped to the same set of proteins represented by the DNA to DNA  mapping tests. As anticipated, the protein alignment method consistently mapped a greater number of reads than BWA, Novoalign and Novoalign degenerate. ACCURACY and SENSITIVITY.  

Modification for improved read ORF prediction. 

The ability to functionally map WMS read comes at a cost of taxonomic resolution.  The proportion of reads mapped to the correct (most closely related) reference taxon was X % for the BWA and Native Novalign whereas the correct taxonomic reference was Y for the Degenerate reference mapping using Novalign and Z for the mapping in amino acid space using PALADIN.


  
  
  